geoportailv3/static/js/exclusionmanagerservice.js

```diff
@@ -0,0 +1,185 @@
+
+  if (!goog.isDef(exclusion1)) {
+    return;
+  }
+
+  var layers = map.getLayers().getArray();
+  var len = layers.length;
+  var i;
+  var layer2;
+  var msg;
+
+  var layersToHide = [];
+  for (i = 0; i < len; i++) {
+    layer2 = layers[i];
+    if (layer2 == layer1 || !goog.isDef(layer2.get('metadata')) ||
+        !goog.isDef(layer2.get('metadata')['exclusion'])) {
+      continue;
+    }
+
+    // check exclusion with current baselayer
+    var exclusion2 = layer2.get('metadata')['exclusion'];
+    opacity = layer2.get('opacity');
+    if (this.intersects_(exclusion1, exclusion2) && goog.isNumber(opacity) &&
+        opacity > 0) {
+      // layer to exclude is not the current base layer
+      if (i !== 0) {
+        layersToHide.push(layer2.get('label'));
+        layer2.set('opacity', 0);
+      } else {
+        this.backgroundLayerMgr_.set(map, this.blankLayer_);
+        msg = 'The background layer has been deactivated because the ' +
+            'layer <b>' + layer1.get('label') +
+            '</b> cannot be displayed on top of it';
+        this.notify_(msg);
+      }
+    }
+  }
+  if (layersToHide.length) {
+    msg = 'The layer(s) <b>' + layersToHide.join(',') +
+        '</b> have been deactivated because they cannot be displayed while ' +
+        ' the layer <b>' + layer1.get('label') +
+        '</b> is displayed';
+    this.notify_(msg);
+  }
+};
+
+
+/**
+ * @param {ol.Map} map The OpenLayers map.
+ */
+app.ExclusionManager.prototype.init = function(map) {
+  var layerOpacityListenerKeys = {};
+
+  // listen on layers being added to the map
+  // base layers switch should fire the event as well
+  map.getLayers().on('add', goog.bind(function(e) {
+    var layer = e.element;
+    this.checkForLayerExclusion_(map, layer);
+
+    // listen on opacity change
+    var key = goog.events.listen(layer, 'change:opacity', goog.bind(
+        function(e) {
+          this.checkForLayerExclusion_(map, layer);
+        }, this));
+    layerOpacityListenerKeys[goog.getUid(layer)] = key;
+  }, this));
+
+  // remove any listener on opacity change when layer is removed from map
+  goog.events.listen(map.getLayers(), 'remove', goog.bind(function(e) {
+    var layer = e.element;
+    goog.events.unlistenByKey(layerOpacityListenerKeys[goog.getUid(layer)]);
+  }, this));
+};
+
+
+/**
+ * The factory function creating the appExclusionManager service.
+ * @param {ngeo.BackgroundLayerMgr} ngeoBackgroundLayerMgr Background layer
+ *     manager.
+ * @param {app.blankLayer} appBlankLayer Blank Layer service.
+ *     app.blankLayer
+ * @param {app.Notify} appNotify Notify service.
+ * @return {app.ExclusionManager} The exclusion manager service.
+ * @private
+ * @ngInject
+ */
+app.ExclusionManagerFactory_ = function(ngeoBackgroundLayerMgr, appBlankLayer,
+    appNotify) {
+  return new app.ExclusionManager(ngeoBackgroundLayerMgr, appBlankLayer,
+      appNotify);
+};
+
+app.module.factory('appExclusionManager', app.ExclusionManagerFactory_);
```

Comment:
Can't you use `module.service` here? What's the role of the intermediate factory?